crates/uv-installer/src/compile.rs

```diff
@@ -0,0 +1,222 @@
+use std::io;
+use std::num::NonZeroUsize;
+use std::path::{Path, PathBuf};
+use std::process::Stdio;
+use std::time::Duration;
+
+use async_channel::{Receiver, SendError};
+use tempfile::tempdir;
+use thiserror::Error;
+use tokio::io::{AsyncBufReadExt, AsyncWriteExt, BufReader};
+use tokio::process::{ChildStdin, ChildStdout, Command};
+use tokio::task::JoinError;
+use tracing::instrument;
+use walkdir::WalkDir;
+
+use uv_warnings::warn_user;
+
+const COMPILEALL_SCRIPT: &str = include_str!("pip_compileall.py");
+const MAIN_TIMEOUT: Duration = Duration::from_secs(10);
+
+#[derive(Debug, Error)]
+pub enum CompileError {
+    #[error("Failed to list files in site-packages")]
+    Walkdir(#[from] walkdir::Error),
+    #[error("Failed to send task to worker")]
+    WorkerDisappeared(SendError<PathBuf>),
+    #[error("The task executor is broken, did some other task panic?")]
+    Join(#[from] JoinError),
+    #[error("Failed to start Python interpreter to run compile script")]
+    PythonSubcommand(#[source] io::Error),
+    #[error("Failed to create temporary script file")]
+    TempFile(#[source] io::Error),
+    #[error("Bytecode compilation failed, expected {0:?}, got: {1:?}")]
+    WrongPath(String, String),
+    #[error("Failed to write to Python stdin")]
+    ChildStdin(#[source] io::Error),
+    #[error("Failed to read Python stdout")]
+    ChildStdout(#[source] io::Error),
+    #[error("Bytecode timed out ({}s)", _0.as_secs_f32())]
+    Timeout(Duration),
+}
+
+/// Bytecode compile all file in `dir` using a pool of work-stealing python interpreters running a
+/// Python script that calls `compileall.compile_file`.
+///
+/// All compilation errors are muted (like pip). There is a 10s timeout to handle the case that
+/// the workers have gotten stuck; This happens way to easily with channels and subprocesses, e.g.
+/// because some pipe is full, we're waiting when it's buffered, or we didn't handle channel closing
+/// properly.
+///
+/// We only compile all files, but we don't update the RECORD, relying on PEP 491:
+/// > Uninstallers should be smart enough to remove .pyc even if it is not mentioned in RECORD.
```

Comment:
Our installer always removes `__pycache__`, so this should be ok, right?

Reply:
Yes, i tested with django and it got removed without residues.